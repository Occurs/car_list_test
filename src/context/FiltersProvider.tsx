import React, { useState, useEffect, createContext } from "react";
import { getManufacturers, getColors } from "client/http";
import { IChildren, IManufacturer } from "types/types";

type IContext = { colors: Array<string>; manufacturers: Array<IManufacturer> };
export const FiltersContext = createContext({} as IContext);

export const FiltersProvider = ({ children }: IChildren) => {
  const [colors, setColors] = useState<Array<string>>([]);
  const [manufacturers, setManufacturers] = useState<Array<IManufacturer>>([]);
  useEffect(() => {
    async function fetchFiltersOptions() {
      const [colors, manufacturers] = await Promise.all([
        getColors(),
        getManufacturers(),
      ]);
      setColors(colors);
      setManufacturers(manufacturers);
    }
    fetchFiltersOptions();
  }, []);
  return (
    <FiltersContext.Provider value={{ colors, manufacturers }}>
      {children}
    </FiltersContext.Provider>
  );
};
