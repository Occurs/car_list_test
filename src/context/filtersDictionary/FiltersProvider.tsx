import React,{ useState, useEffect, createContext } from 'react';
import { getManufacturers, getColors } from 'client/http'
import { IChildren, IManufacturer } from 'types/types';

type IContext = { colors: Array<string>, manufacturers: Array<IManufacturer> }
export const FiltersContext = createContext({} as IContext);

export const FiltersProvider = ({ children }: IChildren) => {
    const [colors, setColors] = useState<Array<string>>([]);
    const [manufacturers, setManufacturers] = useState<Array<IManufacturer>>([]);
    useEffect(() => {
        (async () => {
            const [colors, manufacturers] = await Promise.all([getColors(), getManufacturers()]);
            setColors(colors);
            setManufacturers(manufacturers);
          })();
        }, []);
    return (
    <FiltersContext.Provider value={{ colors, manufacturers }}>
        {children}
    </FiltersContext.Provider>
    );
}