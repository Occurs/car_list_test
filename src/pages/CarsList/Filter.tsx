import React, { useContext } from "react";
import Box from "@material-ui/core/Box";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core/styles";
import MenuItem from "@material-ui/core/MenuItem";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";
import Button from "@material-ui/core/Button";
import { FiltersContext } from "context/filtersDictionary/FiltersProvider";
import { colors as colorPallette } from "styles/variables";

type IFilterComponent = {
  applyFilters: () => void;
  onFiltersHandle:
    | ((
        event: React.ChangeEvent<{
          name?: string | undefined;
          value: unknown;
        }>,
        child: React.ReactNode
      ) => void)
    | undefined;
  color: string;
  manufacturer: string;
};

const useStyles = makeStyles(() => ({
  formControl: {
    marginBottom: "24px",
  },
  button: {
    marginLeft: "auto",
    marginRight: "0px",
    display: "block",
    color: "#FFFFFF !important",
    width: "128px",
  },
}));

const MenuProps = {
  PaperProps: {
    style: {
      borderTop: "none",
      borderRdius: "2px",
      width: "inherit",
      marginTop: "47px",
    },
  },
};

const Filter = ({
  applyFilters,
  onFiltersHandle,
  color,
  manufacturer,
}: IFilterComponent) => {
  const classes = useStyles();
  const { colors, manufacturers } = useContext(FiltersContext);

  return (
    <Box
      component="div"
      border={`2px solid ${colorPallette.secondary}`}
      padding="24px"
      width="100%"
    >
      <Typography>
        <Box component="span" fontSize="12px">
          Color
        </Box>
      </Typography>
      <FormControl
        className={classes.formControl}
        variant="outlined"
        size="small"
        fullWidth
      >
        <Select
          value={color}
          onChange={onFiltersHandle}
          inputProps={{
            name: "color",
            id: "color-select",
          }}
          MenuProps={MenuProps}
          variant="outlined"
          displayEmpty
        >
          <MenuItem key="none" value="">
            All car colors
          </MenuItem>
          {colors.map((color) => (
            <MenuItem key={color} value={color}>
              {color}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
      <Typography>
        <Box component="span" fontSize="12px">
          Manufacturer
        </Box>
      </Typography>
      <FormControl
        className={classes.formControl}
        variant="outlined"
        size="small"
        fullWidth
      >
        <Select
          value={manufacturer}
          onChange={onFiltersHandle}
          inputProps={{
            name: "manufacturer",
            id: "manufacturer-select",
          }}
          MenuProps={MenuProps}
          variant="outlined"
          displayEmpty
        >
          <MenuItem key="none" value="">
            All manufacturers
          </MenuItem>
          {manufacturers.map((manufacturer) => (
            <MenuItem key={manufacturer.name} value={manufacturer.name}>
              {manufacturer.name}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
      <Button
        variant="contained"
        color="primary"
        onClick={applyFilters}
        className={classes.button}
      >
        Filter
      </Button>
    </Box>
  );
};

export default Filter;
