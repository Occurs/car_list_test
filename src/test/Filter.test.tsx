import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, getByTestId } from "@testing-library/react";
import Filter from "pages/CarsList/Filter";
import { FiltersContext } from "context/FiltersProvider";

describe("Filter Component", () => {
  const props = {
    applyFilters: jest.fn(),
  };

  const mock = {
    colors: ["white", "red"],
    manufacturers: [
      {
        name: "Fiat",
        models: [
          {
            name: "Marea",
          },
        ],
      },
      {
        name: "Opel",
        models: [
          {
            name: "Astra",
          },
        ],
      },
    ],
  };

  test("snapshot", () => {
    const component = renderer.create(
      <FiltersContext.Provider value={mock}>
        <Filter {...props} />
      </FiltersContext.Provider>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("click on last page button", () => {
    const { getByText } = render(
      <FiltersContext.Provider value={mock}>
        <Filter {...props} />
      </FiltersContext.Provider>
    );
    const filter = getByText("Filter");
    fireEvent.click(filter);
    expect(props.applyFilters).toHaveBeenCalledTimes(1);
    expect(props.applyFilters).toHaveBeenCalledWith({
      color: "",
      manufacturer: "",
    });
  });
});
