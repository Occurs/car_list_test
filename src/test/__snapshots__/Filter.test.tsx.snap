// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Component snapshot 1`] = `
<div
  className="MuiBox-root MuiBox-root-3"
>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    <span
      className="MuiBox-root MuiBox-root-4"
    >
      Color
    </span>
  </p>
  <div
    className="MuiFormControl-root makeStyles-formControl-1 MuiFormControl-fullWidth"
  >
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="mui-component-select-color"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
        id="mui-component-select-color"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        All car colors
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        id="color-select"
        name="color"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value=""
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
        style={
          Object {
            "paddingLeft": 8,
          }
        }
      >
        <legend
          className="PrivateNotchedOutline-legend-6"
          style={
            Object {
              "width": 0.01,
            }
          }
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </legend>
      </fieldset>
    </div>
  </div>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    <span
      className="MuiBox-root MuiBox-root-9"
    >
      Manufacturer
    </span>
  </p>
  <div
    className="MuiFormControl-root makeStyles-formControl-1 MuiFormControl-fullWidth"
  >
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="mui-component-select-manufacturer"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
        id="mui-component-select-manufacturer"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        All manufacturers
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        id="manufacturer-select"
        name="manufacturer"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value=""
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
        style={
          Object {
            "paddingLeft": 8,
          }
        }
      >
        <legend
          className="PrivateNotchedOutline-legend-6"
          style={
            Object {
              "width": 0.01,
            }
          }
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-2 MuiButton-containedPrimary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Filter
    </span>
  </button>
</div>
`;
